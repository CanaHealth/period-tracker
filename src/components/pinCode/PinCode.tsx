import React from 'react';
import { useState } from 'react';
import { HiBadgeCheck } from 'react-icons/hi';

import clsxm from '@/lib/clsxm';

import BigButton from '@/components/period/calendar/options/BigButton';
import PinInput from '@/components/pinCode/PinInput';

import { saveDataOnChain } from '@/util/saveOnChain';
import {
  createSolanaWallet,
  decryptWallet,
  encryptWallet,
  getPasscodeFromCookie,
  getWalletFromLocalStorage,
  solanaWallet,
  storePasscodeAsCookie,
  storeWalletInLocalStorage,
} from '@/util/WalletOperations';
import { usableSecretKey } from '@/util/WalletOperations';

import AcceptModal from './AcceptModal';

type PinCodeProps = {
  pincode: number[];
  className?: string;
  variant?: 'row' | 'col';
  setPublicKey: React.Dispatch<React.SetStateAction<string>>;
} & React.ComponentPropsWithoutRef<'div'>;

const PinCode: React.FC<PinCodeProps> = ({
  className,
  pincode,
  variant = 'row',
  setPublicKey,
}) => {
  const [loading, setLoading] = useState(false);

  const [pin, setPin] = useState<number[]>(pincode);
  const [open, setOpen] = useState(false);
  const [refIndex, setRefIndex] = useState<number>(0);
  const [nftID, setNftID] = useState<string>('');
  const [blockExplorer, setBlockExplorer] = useState<boolean>(false);

  const onChangeDigits = (value: number, index: number) => {
    setPin(pin.map((digit, i) => (i === index ? value : digit)));
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    const { key } = event;

    switch (key) {
      case 'ArrowLeft':
        setRefIndex(Math.max(0, refIndex - 1));
        break;
      case 'ArrowRight':
        setRefIndex(Math.min(pin.length - 1, refIndex + 1));
        break;
      case '0':
      case '1':
      case '2':
      case '3':
      case '4':
      case '5':
      case '6':
      case '7':
      case '8':
      case '9':
        onChangeDigits(Number(key), refIndex);
        setRefIndex(Math.min(pin.length - 1, refIndex + 1));
        break;

      default:
        break;
    }
  };

  const launchBlockExplorer = () => {
    const url = `https://solscan.io/account/${nftID}`;
    window.open(url, '_blank');
  };

  const submitPinCode = () => {
    const pinConcat = pin.join(''); // Concatenated pincode as string

    if (pin.length === 6) {
      if (getWalletFromLocalStorage().publicKey === '') {
        const wallet = createSolanaWallet();

        // Pass in pin in string format to encrypt wallet.
        const encryptedWallet: solanaWallet = encryptWallet(wallet, pinConcat);

        setPublicKey(encryptedWallet.publicKey);

        // Store wallet in local storage.
        storeWalletInLocalStorage(encryptedWallet);

        storePasscodeAsCookie(pinConcat);
      } else {
        const encryptedWallet = getWalletFromLocalStorage();

        const decryptedWallet: solanaWallet = decryptWallet(
          encryptedWallet,
          pinConcat
        );

        setPublicKey(encryptedWallet.publicKey);

        storePasscodeAsCookie(pinConcat);
      }
    }
  };

  const handleAccept = async () => {
    setLoading(true);
    setBlockExplorer(false);

    // console.log(loading);
    const encryptedWallet = getWalletFromLocalStorage();
    const decryptedWallet: solanaWallet = decryptWallet(
      encryptedWallet,
      getPasscodeFromCookie()
    );

    // TODO: Handle expired cookie.

    saveDataOnChain(
      localStorage.getItem('FLOWDATA'),
      usableSecretKey(decryptedWallet.secretKey)
    )
      .then((id) => {
        setLoading(false);
        setNftID(id);
        setBlockExplorer(true);
      })
      .catch((err) => {
        console.log(err);
        setLoading(false);
      });
  };

  return (
    <div
      className={clsxm(
        'relative mx-auto flex max-w-min justify-center transition-all',
        ' active:animate-blob  ',
        className
      )}
      onKeyDown={handleKeyDown}
    >
      <div
        className={clsxm(
          'flex flex-col border',
          'z-10 mx-auto flex',
          variant === 'row' && ['flex-row']
        )}
      >
        <AcceptModal
          open={open}
          setOpen={setOpen}
          handleSubmit={handleAccept}
          loading={loading}
          blockExplorer={blockExplorer}
          launchBlockExplorer={launchBlockExplorer}
        />

        <div className='absolute inset-x-10 -top-10 -z-10 flex h-12  flex-col items-center justify-center rounded-t-full border-x  border-t bg-white py-1 px-2 text-xs'>
          {' '}
          enter pin:
        </div>

        <div className='mx-auto flex flex-row justify-center'>
          {pin.map((digit, index) => (
            <PinInput
              key={index}
              value={digit}
              refIndex={index}
              myRef={index === refIndex}
              whenFocused={() => setRefIndex(index)}
              className=''
            />
          ))}
        </div>

        <BigButton
          OnClickDo={() => {
            submitPinCode();
            setOpen(true);
          }}
          icon={<HiBadgeCheck />}
          text='Save on-chain'
          iconLocation='r'
          height='20'
          className={clsxm([
            variant === 'row' && ['w-fit px-10'],
            variant === 'col' && ['w-full'],
          ])}
        />
      </div>
    </div>
  );
};

export default PinCode;
