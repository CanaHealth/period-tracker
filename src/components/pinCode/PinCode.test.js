/* eslint-disable unused-imports/no-unused-imports */
import { describe, expect, it, screen } from '@jest/globals';
import { fireEvent, render } from '@testing-library/react';
import { axe } from 'jest-axe';

import PinCode from '@/components/pinCode/PinCode';

describe('PinCode', () => {
  it('should render', () => {
    const { container } = render(
      <PinCode className='' pincode={[0, 0, 0, 0, 0, 0]} />
    );
    expect(container).toMatchSnapshot();
  });
});
