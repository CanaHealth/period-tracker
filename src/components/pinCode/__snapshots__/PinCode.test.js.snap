// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PinCode should render 1`] = `
<div>
  <div
    class="relative mx-auto flex max-w-min justify-center transition-all active:animate-blob"
  >
    <div
      class="border z-10 mx-auto flex flex-row"
    >
      <div
        class="absolute inset-x-10 -top-10 -z-10 flex h-12 flex-col items-center justify-center rounded-t-full border-x border-t bg-white py-1 px-2 text-xs"
      >
         
        enter pin:
      </div>
      <div
        class="mx-auto flex flex-row justify-center"
      >
        <div
          class=" relative"
        >
          <input
            class="appearance-none border-gray-100 text-center text-gray-700 shadow-sm transition-all duration-150 ease-in-out focus:outline-none h-10 w-10 text-xs hover:bg-gray-100 hover:text-gray-900"
            type="number"
            value="0"
          />
        </div>
        <div
          class=" relative"
        >
          <input
            class="appearance-none border-gray-100 text-center text-gray-700 shadow-sm transition-all duration-150 ease-in-out focus:outline-none h-10 w-10 text-xs hover:bg-gray-100 hover:text-gray-900"
            type="number"
            value="0"
          />
        </div>
        <div
          class=" relative"
        >
          <input
            class="appearance-none border-gray-100 text-center text-gray-700 shadow-sm transition-all duration-150 ease-in-out focus:outline-none h-10 w-10 text-xs hover:bg-gray-100 hover:text-gray-900"
            type="number"
            value="0"
          />
        </div>
        <div
          class=" relative"
        >
          <input
            class="appearance-none border-gray-100 text-center text-gray-700 shadow-sm transition-all duration-150 ease-in-out focus:outline-none h-10 w-10 text-xs hover:bg-gray-100 hover:text-gray-900"
            type="number"
            value="0"
          />
        </div>
        <div
          class=" relative"
        >
          <input
            class="appearance-none border-gray-100 text-center text-gray-700 shadow-sm transition-all duration-150 ease-in-out focus:outline-none h-10 w-10 text-xs hover:bg-gray-100 hover:text-gray-900"
            type="number"
            value="0"
          />
        </div>
        <div
          class=" relative"
        >
          <input
            class="appearance-none border-gray-100 text-center text-gray-700 shadow-sm transition-all duration-150 ease-in-out focus:outline-none h-10 w-10 text-xs hover:bg-gray-100 hover:text-gray-900"
            type="number"
            value="0"
          />
        </div>
      </div>
      <button
        class="flex w-44 flex-col items-center justify-center rounded shadow-sm hover:shadow-md border border-blue-93 bg-white text-indigo-800 transition-all ease-out active:bg-gray-800 active:text-white hover:bg-gray-600 hover:text-white h-10 text-xl w-fit px-10"
        type="button"
      >
        <div
          class="flex items-center justify-center text-inherit flex-row-reverse"
        >
          <span
            class="m-1"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 20 20"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M6.267 3.455a3.066 3.066 0 001.745-.723 3.066 3.066 0 013.976 0 3.066 3.066 0 001.745.723 3.066 3.066 0 012.812 2.812c.051.643.304 1.254.723 1.745a3.066 3.066 0 010 3.976 3.066 3.066 0 00-.723 1.745 3.066 3.066 0 01-2.812 2.812 3.066 3.066 0 00-1.745.723 3.066 3.066 0 01-3.976 0 3.066 3.066 0 00-1.745-.723 3.066 3.066 0 01-2.812-2.812 3.066 3.066 0 00-.723-1.745 3.066 3.066 0 010-3.976 3.066 3.066 0 00.723-1.745 3.066 3.066 0 012.812-2.812zm7.44 5.252a1 1 0 00-1.414-1.414L9 10.586 7.707 9.293a1 1 0 00-1.414 1.414l2 2a1 1 0 001.414 0l4-4z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </div>
      </button>
    </div>
  </div>
</div>
`;
